bpf/kmesh/workload/include/authz.h

```diff
@@ -32,79 +49,187 @@ static inline wl_policies_v *get_workload_policies_by_uid(__u32 workload_uid)
 #pragma unroll
         for (i = 0; i < MAX_MEMBER_NUM_PER_POLICY; i++) {
             if (i >= match->n_not_destination_ports) {
                 break;
             }
-            if (info->iph->version == 4) {
-                if (bpf_htons(notPorts[i]) == tuple_info->ipv4.dport) {
-                    return UNMATCHED;
-                }
-            } else {
-                if (bpf_htons(notPorts[i]) == tuple_info->ipv6.dport) {
-                    return UNMATCHED;
+            if (bpf_htons(notPorts[i]) == dport) {
+                if (res->action == AUTH_DENY) {
+                    return XDP_PASS;
+                } else {
+                    BPF_LOG(INFO, AUTH, "Denied: dport %u matches \n", notPorts[i]);
+                    return XDP_DROP;
                 }
             }
         }
     }
-    // if not match not_destination_ports && has no destination_ports, return MATCHED
+
     if (match->n_destination_ports == 0) {
-        return MATCHED;
+        bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
+        return (res->action == AUTH_DENY) ? XDP_DROP : XDP_PASS;
     }
 
     ports = kmesh_get_ptr_val(match->destination_ports);
     if (!ports) {
-        return UNMATCHED;
+        return (res->action == AUTH_DENY) ? XDP_PASS : XDP_DROP;
     }
+
 #pragma unroll
     for (i = 0; i < MAX_MEMBER_NUM_PER_POLICY; i++) {
         if (i >= match->n_destination_ports) {
             break;
         }
-        if (info->iph->version == 4) {
-            if (bpf_htons(ports[i]) == tuple_info->ipv4.dport) {
-                return MATCHED;
-            }
-        } else {
-            if (bpf_htons(ports[i]) == tuple_info->ipv6.dport) {
-                return MATCHED;
+        if (bpf_htons(ports[i]) == dport) {
+            if (res->action == AUTH_DENY) {
+                BPF_LOG(INFO, AUTH, "Denied: dport %u matches \n", ports[i]);
+                return XDP_DROP;
+            } else {
+                return XDP_PASS;
             }
         }
     }
-    return UNMATCHED;
+
+    bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
+    return XDP_PASS;
 }
 
-static inline int match_check(Istio__Security__Match *match, struct xdp_info *info, struct bpf_sock_tuple *tuple_info)
+static inline int match_check(struct xdp_md *ctx, void *match, Istio__Security__Action action)
 {
-    __u32 matchResult;
+    struct match_ctx res;
+    struct bpf_sock_tuple tuple_key = {0};
+    struct xdp_info info = {0};
+    int ret;
+
+    if (get_tuple_key(ctx, &tuple_key, &info) != PARSER_SUCC) {
+        BPF_LOG(ERR, AUTH, "Failed to get tuple key\n");
+        return XDP_ABORTED;
+    }
+    res.match = match;
+    res.action = action;
 
-    // if multiple types are set, they are AND-ed, all matched is a match
-    // todo: add other match types
-    matchResult = matchDstPorts(match, info, tuple_info);
-    return matchResult;
+    ret = bpf_map_update_elem(&tailcall_info_map, &tuple_key, &res, BPF_ANY);
```

Comment:
If there are multi policy set, multiple packet concurrently received, isn't this map overwritten with random policy?